lib/systemd/user-level-services.nix

```diff
@@ -0,0 +1,105 @@
+{ nixpkgs
+, system ? "x86_64-linux"
+, pkgs ? nixpkgs.legacyPackages.${system}, ... }:
+let
+  mkVM = moduleConfig: serviceConfig: nixpkgs.lib.nixosSystem {
+    inherit system;
+    modules = [ ({ ... }: {
+
+      imports = [ moduleConfig.module ];
+
+      services.${moduleConfig.nixOsServiceName} = {
+        enable = true;
+      } // serviceConfig;
+
+    }) ];
+  };
+
+in {
+  mkActivateScript = activateScriptName:
+      { module, nixOsServiceName, primaryService, primaryPortOptionName, auxiliaryServices}@moduleConfig:
+      { backupAction ? "", checkAction ? null, restoreAction ? ""}:
+      serviceConfig:
+    let
+      vm = mkVM moduleConfig serviceConfig;
+      allServices = [primaryService] ++ auxiliaryServices;
+    in
+    pkgs.writeShellScriptBin activateScriptName ''
+      set -euo pipefail
+      export XDG_RUNTIME_DIR="/run/user/$UID"
+      mkdir -v -p "$HOME/.config/systemd/user/default.target.wants"
+      ${if backupAction != ""
+      then
+        ''
+          # Deploy may change DB schema via migration and we should be able
+          # to restore its state to a point before an update to successfully rollback
+          # failed deployment
+          echo Stopping ${primaryService}
+          systemctl --user stop ${primaryService} || true
+          echo Backing up the existing state
+          ${backupAction}
+        ''
+      else
+        ""
+      }
+
+      ${pkgs.lib.concatStringsSep "\n" (builtins.map (service:
+        let
+          serviceUnit = vm.config.systemd.units.${service};
+        in
+          ''
+            rm -v -f -- "$HOME/.config/systemd/user/${service}"
+            ln -v -s ${serviceUnit.unit}/${service} "$HOME/.config/systemd/user/${service}"
+          ''
+          +
+          (if builtins.elem "default.target" serviceUnit.wantedBy
+          then
+            ''
+              rm -v -f -- "$HOME/.config/systemd/user/default.target.wants/${service}"
+              ln -v -s "$HOME/.config/systemd/user/${service}" "$HOME/.config/systemd/user/default.target.wants"
```

Comment:
Is there a reason why `default.target` is hardcoded? AFAIU, you can simply `ln -v -s "$HOME/.config/systemd/user/${service}" "$HOME/.config/systemd/user/${serviceUnit.wantedBy}.wants"` for each value of `serviceUnit.wantedBy`

Reply:
changed